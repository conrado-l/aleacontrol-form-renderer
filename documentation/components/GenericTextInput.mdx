---
name: GenericTextInput
---

import { Playground, PropsTable } from 'docz'
import GenericTextInput from '../../src/components/GenericTextInput.js'

# GenericTextInput

<PropsTable of={GenericTextInput} />

## Description
An input that can be either plain text or email depending on the component prop
## Demo and usage

<Playground>
    <GenericTextInput component='string' name='text-input' value='Hello' col={6} update={(e) => e}/>
</Playground>

<Playground>
    <GenericTextInput component='email' name='email-input' value='info@aleacontrol.com' col={6} update={(e) => e}/>
</Playground>
