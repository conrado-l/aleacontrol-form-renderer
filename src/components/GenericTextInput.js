import React, { Component } from 'react'
import PropTypes from 'prop-types'
import InputTypes from '../consts/inputTypes'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Form from 'react-bootstrap/Form'

/**
 * Generic input that can be used for email or text. It can support more string types, adding them to the mapping.
 */
class GenericTextInput extends Component {
  constructor (props) {
    super(props)
    this.state = {
      value: props.value
    }

    this.handleChange = this.handleChange.bind(this)
  }

  /**
     * Updates the input value and sends the value through the parent callback function
     * @param {object} e Event
     */
  handleChange (e) {
    const newValue = e.target.value
    this.setState({ value: newValue })
    this.props.update(this.props.name, newValue)
  }

    /**
     * Mapping: component name -> input type
     */
    types = {
      [InputTypes.STRING]: 'text',
      [InputTypes.EMAIL]: 'email'
    }

    /**
     * Gets the input type based on the types mapping
     * @returns {string}
     */
    inputType () {
      return this.types[this.props.component]
    }

    render () {
      return (
        <Row>
          <Col xs={12} sm={this.props.col}>
            <Form.control type={this.inputType()}
              name={this.props.name}
              value={this.state.value || ''}
              onChange={this.handleChange} />
          </Col>
        </Row>
      )
    }
}

GenericTextInput.propTypes = {
  component: PropTypes.string,
  name: PropTypes.string,
  value: PropTypes.string,
  col: PropTypes.number,
  update: PropTypes.func
}
export default GenericTextInput
