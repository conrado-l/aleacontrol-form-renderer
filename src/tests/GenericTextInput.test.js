import React from 'react'
import { mount } from 'enzyme'
import GenericTextInput from '../components/EmailInput'
import renderer from 'react-test-renderer'

describe('GenericTextInput', () => {
  const mockupEmailProps = {
    name: 'email',
    value: 'paul@gmail.com'
  }
  const mockupTextProps = {
    name: 'first_name',
    value: 'Paul'
  }

  it('should render correctly generic text input component', () => {
    const inputFn = jest.fn()
    const GenericTextInputComponent = renderer
      .create(<GenericTextInput component='email' name={mockupEmailProps.name} value={mockupEmailProps.value}
        update={inputFn} />)
      .toJSON()
    expect(GenericTextInputComponent).toMatchSnapshot()
  })

  it('should render the email input with the proper value and name', () => {
    const inputFn = jest.fn()
    const GenericTextInputComponent = mount(<GenericTextInput component='email' name={mockupEmailProps.name}
      value={mockupEmailProps.value} update={inputFn} />)
    expect(
      GenericTextInputComponent.containsMatchingElement(
        <input type='email' value={mockupEmailProps.value} name={mockupEmailProps.name} />
      )
    ).toBeTruthy()
  })

  it('should render the text input with the proper value and name', () => {
    const inputFn = jest.fn()
    const GenericTextInputComponent = mount(<GenericTextInput component='string' name={mockupTextProps.name}
      value={mockupTextProps.value} update={inputFn} />)

    expect(
      GenericTextInputComponent.containsMatchingElement(
        <input name={mockupTextProps.name} value={mockupTextProps.value} />
      )
    ).toBeTruthy()
  })

  it('should call the update callback function with proper parameters on input', () => {
    const inputFn = jest.fn()
    const GenericTextInputComponent = mount(<GenericTextInput name={mockupTextProps.name}
      value={mockupTextProps.value} update={inputFn} />)
    const inputString = 'Aleacontrol'

    GenericTextInputComponent.find('input').simulate('change', {
      target: { value: inputString }
    })

    expect(inputFn).toHaveBeenCalledWith(mockupTextProps.name, inputString)
  })
})
