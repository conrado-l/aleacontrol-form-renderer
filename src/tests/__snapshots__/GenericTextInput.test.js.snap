// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericTextInput should render correctly generic text input component 1`] = `
<div
  className="row"
>
  <div
    className="col-12"
  >
    <input
      className="form-control"
      name="email"
      onChange={[Function]}
      type="email"
      value="paul@gmail.com"
    />
  </div>
</div>
`;
